### PR TITLE
Update S2328: "GetHashCode should not reference mutable fields" should report once per method

### LINKED ISSUE #346: Update S2328: "GetHashCode should not reference mutable fields" should report once per method
[RSPEC-2328](https://jira.sonarsource.com/browse/RSPEC-2328)

All currently raised issues in the method should be converted to secondary locations.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #346.

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/GetHashCodeMutable.cs [hunk 1]
```diff
@@ -74,15 +73,25 @@ protected sealed override void Initialize(SonarAnalysisContext context)
                         .Where(symbol => symbol != null)
                         .ToImmutableHashSet();
 
-                    var identifiers = methodSyntax.DescendantNodes()
-                        .OfType<IdentifierNameSyntax>();
+                    var identifiers = methodSyntax.DescendantNodes().OfType<IdentifierNameSyntax>();
+
+                    var secondaryLocations = GetAllFirstMutableFieldsUsed(c, fieldsOfClass, identifiers)
+                        .Select(identifierSyntax => new SecondaryLocation(identifierSyntax.GetLocation(),
+                            string.Format(SecondaryMessageFormat, identifierSyntax.Identifier.Text)))
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/GetHashCodeMutable.cs]: This Select is a bit ugly and could be extracted into a method...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/GetHashCodeMutableCodeFixProvider.cs [hunk 2]
```diff
@@ -49,41 +52,65 @@ public sealed override FixAllProvider GetFixAllProvider()
...
+                .Where(idenfitierName => idenfitierName != null)
+                .ToList();
+
+            if (identifiersToFix.Count == 0)
             {
                 return;
             }
 
             var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+            var allFieldDeclarationTasks = identifiersToFix.Select(identifier => GetFieldDeclarationSyntax(semanticModel, identifier, context.CancellationToken));
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/GetHashCodeMutableCodeFixProvider.cs]: We should reduce our "line too long" rule limit to 130 chars. GitHub is showing scrollbar for longer lines...

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Verifier.cs [hunk 3]
```diff
@@ -300,19 +300,33 @@ private static void VerifyNoIssueReported(string path, string assemblyName, Diag
...
 
+        private static string ReplaceNonCompliantComment(string line)
+        {
+            var match = Regex.Match(line, IssueLocationCollector.ISSUE_LOCATION_PATTERN);
+            if (!match.Success)
+            {
+                return line;
+            }
+
+            if (match.Groups["issueType"].Value.Equals("Noncompliant"))
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Verifier.cs]: Change to:
match.Groups["issueType"].Value == "Noncompliant"

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/KnownMethods.cs [hunk 4]
```diff
@@ -50,7 +50,8 @@ public static bool IsObjectEquals(this IMethodSymbol methodSymbol)
 
         public static bool IsObjectGetHashCode(this IMethodSymbol methodSymbol)
         {
-            return methodSymbol.IsOverride &&
+            return methodSymbol != null &&
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/KnownMethods.cs]: Good idea. We should do it for all other methods in this class.

Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/KnownMethods.cs]: @michalb-sonar Yes I think so too but this refactoring doesn't belong to this PR that's why I haven't done it yet.
